stdlib/format.ml

```diff
@@ -352,77 +309,67 @@ let format_pp_token state size = function
     if insertion_point > state.pp_max_indent then
       (* can not open a box right there. *)
       begin pp_force_break_line state end;
-    let offset = state.pp_space_left - off in
-    let bl_type =
-      begin match ty with
+    let width = state.pp_space_left - off in
+    let box_type =
+      match ty with
       | Pp_vbox -> Pp_vbox
       | Pp_hbox | Pp_hvbox | Pp_hovbox | Pp_box | Pp_fits ->
-        if size > state.pp_space_left then ty else Pp_fits
-      end in
-    state.pp_format_stack <-
-      Format_elem (bl_type, offset) :: state.pp_format_stack
+        if size > state.pp_space_left then ty else Pp_fits in
+    Stack.push { box_type; width } state.pp_format_stack
 
   | Pp_end ->
-    begin match state.pp_format_stack with
-    | _ :: ls -> state.pp_format_stack <- ls
-    | [] -> () (* No more box to close. *)
-    end
+    Stack.pop_opt state.pp_format_stack |> ignore
 
   | Pp_tbegin (Pp_tbox _ as tbox) ->
-    state.pp_tbox_stack <- tbox :: state.pp_tbox_stack
+    Stack.push tbox state.pp_tbox_stack
 
   | Pp_tend ->
-    begin match state.pp_tbox_stack with
-    | _ :: ls -> state.pp_tbox_stack <- ls
-    | [] -> () (* No more tabulation box to close. *)
-    end
+    Stack.pop_opt state.pp_tbox_stack |> ignore
 
   | Pp_stab ->
-    begin match state.pp_tbox_stack with
-    | Pp_tbox tabs :: _ ->
+    begin match Stack.top_opt state.pp_tbox_stack with
+    | Some (Pp_tbox tabs) ->
       let rec add_tab n = function
         | [] -> [n]
         | x :: l as ls -> if n < x then n :: ls else x :: add_tab n l in
       tabs := add_tab (state.pp_margin - state.pp_space_left) !tabs
-    | [] -> () (* No open tabulation box. *)
+    | None -> () (* No open tabulation box. *)
```

Comment:
Minor nitpick: I tend to think that a conditional or match is more readable when the shortest case is handled first, so I would have the `None` clause first in these cases.